src/tls/server.c

```diff
@@ -0,0 +1,902 @@
+          remove_connection (fd, NULL);
+          return;
+        }
+
+      debug ("TLS handshake completed");
+
+      connection_set_tls_session (con, session);
+    }
+
+  connection_init_ws (con);
+  if (!con->ws)
+    remove_connection (fd, NULL);
+}
+
+static int
+send_all (int socket, const char *data, size_t data_size, int flags)
+{
+  int r;
+
+  while (data_size > 0)
+  {
+    r = send (socket, data, data_size, flags);
+
+    if (r < 0)
+      {
+        if (errno == EAGAIN || errno == EINTR)
+          continue;
+        return -1;
+      }
+    if (r == 0) /* Should Not Happen™, as data_size > 0 */
+      {
+        errno = EIO;
+        return -1;
+      }
+
+    data += r;
+    data_size -= r;
+  }
+  return 0;
+}
+
+static int
+gnutls_send_all (gnutls_session_t session, const char *data, size_t data_size)
+{
+  int r;
+
+  while (data_size > 0)
+  {
+    r = gnutls_record_send (session, data, data_size);
+
+    if (r == GNUTLS_E_AGAIN || r == GNUTLS_E_INTERRUPTED)
+      continue;
+    if (r < 0)
+      return r;
+    if (r == 0) /* Should Not Happen™, as data_size > 0 */
+      return GNUTLS_E_PUSH_ERROR;
+
+    data += r;
+    data_size -= r;
+  }
+  return 0;
+}
+
+/**
+ * handle_connection_data: Handle event on client or ws fd
+ *
+ * We want to avoid any interpretation of data to avoid vulnerabilities, so for
+ * the most part this just means shovelling data between the client and ws. The
+ * only exception is the very first byte of a new connection, to tell apart TLS
+ * from plain HTTP (handled by handle_connection_data_first).
+ */
+static void
+handle_connection_data (int fd)
+{
+  enum DataSource src;
+  Connection *con;
+  int ret;
+  char buffer[1024 * 1024];
+
+  con = get_fd_connection (fd, &src);
+  assert (con);
+  debug ("%s connection fd %i has data from %s; ws %s",
+         con->is_tls ? "TLS" : "unencrypted", con->client_fd,
+         src == WS ? "ws" : "client",
+         con->ws ? con->ws->socket.sun_path : "uninitialized");
+
+  /* first data on a new connection; determine if TLS, init TLS, and assign a ws */
+  if (!con->ws)
+    {
+      assert (src == CLIENT);
+      handle_connection_data_first (fd, con);
+      return;
+    }
+
+  /* TLS connections */
+  if (con->is_tls)
+    {
+      if (src == CLIENT)
+        {
+          TLS_RETRY (ret, gnutls_record_recv (con->session, buffer, sizeof (buffer)));
+          if (ret == 0)
+            {
+              debug ("peer has closed the TLS connection");
+              TLS_RETRY_BLOCK (ret, gnutls_bye (con->session, GNUTLS_SHUT_WR));
+              remove_connection (fd, NULL);
+            }
+          else if (ret < 0)
+            {
+              if (gnutls_error_is_fatal (ret))
+                {
+                  warnx ("%s", gnutls_strerror (ret));
+                  remove_connection (fd, NULL);
+                }
+            }
+          else
+            {
+              int r = send_all (con->ws_fd, buffer, ret, 0);
```

Comment:
This write might block, which could lead to dead-locks with cockpit-ws if cockpit-ws would also do blocking writes.  Cockpit-ws does not block on writes, so we are good.  This should be documented here, however.

I think with pure HTTP, blocking writes on both sides would might be okay, too, since it has strict request/response alternating. But with our WebSocket anything goes.

Reply:
If I understood you correctly, c-tls needs to give up on the "read and corresponding write is one single event handler" approach, and split this up. In particular:

 * each Connection object gets its own data buffer (instead of the single one in server.c's `handle_connection_data()`), and an `event_mode` flag that signifies whether a read() happened recently and the buffer is full (`EPOLLOUT`), or we wait for a read() to happen and the buffer is empty (`EPOLLIN`).
 * server.c splits `handle_connection_data()` into a read and write part. After each read, the epoll gets `EPOLL_CTL_MOD`ed to `EPOLLOUT` (to avoid getting constantly interrupted by `EPOLLINs` while the data isn't flushed out), and it waits until the target fd is ready to write.
 * Once a write of a connection buffer completes, flip the epoll back to `EPOLLIN` and change the connection's `event_mode` back to `EPOLLIN`.
 * The current `send_all()` function has to be given up in favor of adding a buffer offset to Connection, to allow partial writes until the target fd is ready to write more.

Question: Can send() still fail with `EAGAIN` even after epoll sends an `EPOLLOUT` event? That would be counter-intuitive, and at that point there isn't much else that we can do other than "try the send again in the next loop".

Reply:
> [...] c-tls needs to give up on the "read and corresponding write is one single event handler" approach

That would be ideal and necessary for the general case, but the current approach is valid as it is since cockpit-ws is already careful not to block. Cockpit-ws will always make progress and thus cockpit-tls will eventually finish the blocking write.

> In particular:

Exactly right.

> Can send() still fail with EAGAIN even after epoll sends an EPOLLOUT event?

I don't think so since the fd is not in non-blocking mode.

Reply:
> I think that is unlikely, but why would that be a problem?

It was just some edge case that came to my mind. If it happens at all (which would sound like a kernel bug really), it would hopefully clear up in the next iteration. I. e. we don't need to do anything special about this case code-wise.

> put the fd into non-blocking mode

Right, that's the point of it. This needs some thought wrt. the `gnutls_send_all()` case below, but it would just retry a few times on EAGAIN if it happens in non-blocking mode.

Reply:
We discussed to do that as a follow-up. I created https://projects.engineering.redhat.com/browse/COCKPIT-473 to track this.